### ISSUE TITLE
Given up with tabs.

### ISSUE THREAD

**Issue by rdebath** [uncrustify/uncrustify#395]
I've tried to get uncrustify to output the tabs that I want to use, but it appears that it has trouble keeping track (could quite possibly be me having the trouble keeping track!). So right now I've told uncrustify to **never** output tabs and I pipe it out through `unexpand -a`, this gives me what I want.

The problem as I see it is that there are two concepts here; the tab character and the tabstop column position and some configurations aren't sure which they are changing/using.

I think the processing for _tab characters_ should go like this:
* On input expand **all** tabs to spaces at the input_tab_size, no exceptions.
* All processing is done using spaces, **nothing** inserts tabs.
* On output use the output_tab_size to either
    1. Leading spaces are turned into tabs. (The unix `unexpand` command)
    2. Replace all spaces where there will be a reduction in file size. (`unexpand -a`)
    * An output_tab_size of 1 obviously means "never use tabs"

A possible extra option would be an additional flag that says that if the line is unchanged and the input and output tab sizes match and probably several other conditions the input line could be copied precisely to the output. (But that doesn't seem useful to me)

_Tabstop positions_ would have another tabstop_size configuration for allowing the "nudging" of the position that seems to happen sometimes. (I'm assuming that this is actually an intentional "use the next tabstop position" here and not a bug) 

Finally, I notice the `--tabs=LIST` option on `expand` and `unexpand` I suppose someone must have a use for them and I can possibly see such a list being nice for a "use the next tabstop position" configuration. 

In Summary: I currently find the tab output functions unusable. This may be a documentation problem or a confused implementation of some of the "tab" related configuration options. The tab input functionality seems stable but the tab output is too unpredictable to use.

**Comment by bengardner** [comment]
How are you setting `indent_with_tabs`? (spaces, mixed, or tabs?)
I suspect you didn't set that option appropriately.

Uncrustify operates on 'columns' not spaces.  Tabs are expanded to the appropriate column on input.

For modes `indent_with_tabs=0` (spaces) and `indent_with_tabs=2` (tabs), the output is simple.

The output is a bit more complicated for `indent_with_tabs=1`, as it should indent with tabs and align with spaces. In this mode, the tab size in your editor doesn't matter - the text will always look right!

**Comment by rdebath** [comment]
> ... the tab size in your editor doesn't matter - the text will always look right!

Oh, that could be the problem.
You realise for that you're assuming that the tabstop and indent are matched. (Or rather that an indent is a whole number of tabs)

If you want your indent to be 4 and your tabstop to be 8 you will have to have an indent of half a tab -- ie in spaces. Once you do that the editor MUST know how big a tab is, so it's got to be eight as that's the only real standard. If the editor assumes the tabstop was 4 then half your indents will go away as N+1 tabs becomes the same as N+ 0.5*oldTab.

Currently, I'm using "indent_with_tabs=0" this is predictable and seems to work perfectly ie: Yes, uncrustify works with columns. (I then "compress" it "unexpand"). Neither of the other two work for me.

NOTE: When editing, I generally use the "smarttab" mode of VIM. It makes the tab key do indents at the start of the line (of 3 or 4 spaces) which it converts to 8 column tabs, if they match, and tab characters within the line. At all times physical tabs are at a column spacing of 8 characters.
